{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Files\n",
    "\n",
    "Germplasm.tsv and LocusGene.tsv contain the datasets we need for the exam.\n",
    "\n",
    "Our objective is to create a database to contain the data in these files, insert the data into the database, then query the database in a variety of ways.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem 1: Controls\n",
    "\n",
    "Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx). (hint: This will help you decide how to load the data into the database)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Both rows have the same data\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "import re\n",
    "with open(\"Germplasm.tsv\", \"r\") as germplasmfile,\\\n",
    "     open(\"LocusGene.tsv\", \"r\") as locusgenefile:\n",
    "        germplasm = csv.DictReader(germplasmfile, delimiter=\"\\t\", quotechar='\"') #Dictionary reader\n",
    "        locusgene = csv.DictReader(locusgenefile, delimiter=\"\\t\", quotechar='\"')\n",
    "        for row in germplasm:\n",
    "            mo1 = row[\"Locus\"]\n",
    "            #print ('Germplasm locus', mo1)\n",
    "        for row in locusgene:\n",
    "            mo2 = row[\"Locus\"]\n",
    "            #print ('Locus Gene locus', mo2)\n",
    "        if mo1 == mo2:\n",
    "                print (\"Both rows have the same data\")\n",
    "        else:\n",
    "               print (\"No match!!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Both rows have the same data\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "import re\n",
    "with open(\"Germplasm.tsv\", \"r\") as germplasmfile,\\\n",
    "     open(\"LocusGene.tsv\", \"r\") as locusgenefile:\n",
    "        germplasm = csv.DictReader(germplasmfile, delimiter=\"\\t\", quotechar='\"') #Dictionary reader\n",
    "        row1 = next(germplasm)\n",
    "        locusgene = csv.DictReader(locusgenefile, delimiter=\"\\t\", quotechar='\"')\n",
    "        row2 = next(locusgene)\n",
    "        if row1[\"Locus\"] == row2[\"Locus\"]:\n",
    "                print (\"Both rows have the same data\")\n",
    "        else:\n",
    "               print (\"No match!!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem 2:  Design and create the database.  \n",
    "* It should have two tables - one for each of the two data files.\n",
    "* The two tables should be linked in a 1:1 relationship\n",
    "* you may use either sqlMagic or pymysql to build the database\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "8 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>information_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Exam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>LIMS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>mysql</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>performance_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>sys</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>testing123</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('information_schema',),\n",
       " ('Exam',),\n",
       " ('LIMS',),\n",
       " ('germplasm',),\n",
       " ('mysql',),\n",
       " ('performance_schema',),\n",
       " ('sys',),\n",
       " ('testing123',)]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql\n",
    "#%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql\n",
    "\n",
    "%sql show databases;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "2 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "7 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>information_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>LIMS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>mysql</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>performance_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>sys</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>testing123</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('information_schema',),\n",
       " ('LIMS',),\n",
       " ('germplasm',),\n",
       " ('mysql',),\n",
       " ('performance_schema',),\n",
       " ('sys',),\n",
       " ('testing123',)]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#%sql drop database Exam\n",
    "#%sql show databases;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql create database Exam;\n",
    "%sql show databases\n",
    "%sql use Exam;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "5 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Field</th>\n",
       "        <th>Type</th>\n",
       "        <th>Null</th>\n",
       "        <th>Key</th>\n",
       "        <th>Default</th>\n",
       "        <th>Extra</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>id</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td>PRI</td>\n",
       "        <td>None</td>\n",
       "        <td>auto_increment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Locus</td>\n",
       "        <td>varchar(25)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Gene</td>\n",
       "        <td>varchar(20)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ProteinLength</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm_id</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('id', 'int(11)', 'NO', 'PRI', None, 'auto_increment'),\n",
       " ('Locus', 'varchar(25)', 'NO', '', None, ''),\n",
       " ('Gene', 'varchar(20)', 'NO', '', None, ''),\n",
       " ('ProteinLength', 'int(11)', 'NO', '', None, ''),\n",
       " ('germplasm_id', 'int(11)', 'NO', '', None, '')]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql CREATE TABLE germplasm(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, Locus VARCHAR(25) NOT NULL, germplasm VARCHAR(30) NOT NULL, phenotype VARCHAR(2000) NOT NULL, pubmed INTEGER NOT NULL);\n",
    "%sql DESCRIBE germplasm\n",
    "%sql CREATE TABLE locusgene(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, Locus VARCHAR(25) NOT NULL, Gene VARCHAR(20) NOT NULL, ProteinLength INTEGER NOT NULL, germplasm_id INTEGER NOT NULL);\n",
    "%sql DESCRIBE locusgene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Field</th>\n",
       "        <th>Type</th>\n",
       "        <th>Null</th>\n",
       "        <th>Key</th>\n",
       "        <th>Default</th>\n",
       "        <th>Extra</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>id</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td>PRI</td>\n",
       "        <td>None</td>\n",
       "        <td>auto_increment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Locus</td>\n",
       "        <td>varchar(25)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Gene</td>\n",
       "        <td>varchar(20)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>ProteinLength</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm_id</td>\n",
       "        <td>int(11)</td>\n",
       "        <td>NO</td>\n",
       "        <td></td>\n",
       "        <td>None</td>\n",
       "        <td></td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('id', 'int(11)', 'NO', 'PRI', None, 'auto_increment'),\n",
       " ('Locus', 'varchar(25)', 'NO', '', None, ''),\n",
       " ('Gene', 'varchar(20)', 'NO', '', None, ''),\n",
       " ('ProteinLength', 'int(11)', 'NO', '', None, ''),\n",
       " ('germplasm_id', 'int(11)', 'NO', '', None, '')]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql DESCRIBE locusgene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem 3: Fill the database\n",
    "Using pymysql, create a Python script that reads the data from these files, and fills the database.  There are a variety of strategies to accomplish this.  I will give all strategies equal credit - do whichever one you are most confident with."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "import csv \n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Exam',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open('Germplasm.tsv')as csvfile:\n",
    "            germplasm = csv.DictReader(csvfile, delimiter=\"\\t\", quotechar='\"' )\n",
    "            for row in germplasm:\n",
    "                print(row)\n",
    "                sql = \"\"\"INSERT INTO germplasm (Locus, germplasm, phenotype, pubmed ) values \n",
    "                ('{}','{}','{}','{}')\"\"\".format (row[\"Locus\"], row[\"germplasm\"], row[\"phenotype\"], row[\"pubmed\"])\n",
    "                cursor.execute(sql)\n",
    "        \n",
    "        connection.commit()\n",
    "\n",
    "finally: \n",
    "    print (\"\")\n",
    "    connection.close() #If I close before I commit, the inserts are lost"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM germplasm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "import csv \n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Exam',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open('LocusGene.tsv') as csvfile:\n",
    "            germplasm = csv.DictReader(csvfile, delimiter=\"\\t\", quotechar='\"' )\n",
    "            for row in germplasm:\n",
    "                print(row)\n",
    "                locus = row[\"Locus\"]\n",
    "                locus = locus.strip()\n",
    "                print(\"SELECT id FROM germplasm WHERE locus='{}'\".format(locus))\n",
    "                \n",
    "                sql = \"SELECT id FROM germplasm WHERE locus='{}'\".format(locus)\n",
    "                cursor.execute(sql)\n",
    "                results = cursor.fetchall()\n",
    "                germplasmid = results[0]['id']\n",
    "                print (germplasmid)\n",
    "                print (\"Project Id is {}\".format(project))\n",
    "                sql = \"\"\"INSERT INTO locusgene (Locus, Gene, ProteinLength,germplasm_id) values \n",
    "                ('{}','{}','{}','{}')\"\"\".format (row[\"Locus\"], row[\"Gene\"], row[\"ProteinLength\"], germplasmid)\n",
    "                cursor.execute(sql)\n",
    "        \n",
    "        connection.commit()\n",
    "finally: \n",
    "    print (\"\")\n",
    "    connection.close() #If I close before I commit, the inserts are lost"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM locusgene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem 4: Create reports, written to a file\n",
    "\n",
    "1. Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "\n",
    "2. Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "3. Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "4. Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "When creating reports 2 and 3, remember the \"Don't Repeat Yourself\" rule! \n",
    "\n",
    "All reports should be written to **the same file**.  You may name the file anything you wish."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Query 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[{'id': 1, 'Locus': 'AT1G01040', 'germplasm': 'CS3828', 'phenotype': 'Increased abundance of miRNA precursors.', 'pubmed': 17369351, 'locusgene.id': 1, 'locusgene.Locus': 'AT1G01040', 'Gene': 'DCL1', 'ProteinLength': 332, 'germplasm_id': 1}, {'id': 2, 'Locus': 'AT1G01060', 'germplasm': 'lhy-101', 'phenotype': 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 'pubmed': 16891401, 'locusgene.id': 2, 'locusgene.Locus': 'AT1G01060', 'Gene': 'LHY', 'ProteinLength': 290, 'germplasm_id': 2}, {'id': 3, 'Locus': 'AT1G01140', 'germplasm': 'SALK_058629', 'phenotype': 'hypersensitive to low potassium media', 'pubmed': 17486125, 'locusgene.id': 3, 'locusgene.Locus': 'AT1G01140', 'Gene': 'CIPK9', 'ProteinLength': 223, 'germplasm_id': 3}, {'id': 4, 'Locus': 'AT1G01220', 'germplasm': 'SALK_012400C', 'phenotype': 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 'pubmed': 18199744, 'locusgene.id': 4, 'locusgene.Locus': 'AT1G01220', 'Gene': 'FKGP', 'ProteinLength': 190, 'germplasm_id': 4}, {'id': 5, 'Locus': 'AT2G03720', 'germplasm': 'SALK_042433', 'phenotype': 'Multiple straight hairs', 'pubmed': 16367956, 'locusgene.id': 5, 'locusgene.Locus': 'AT2G03720', 'Gene': 'MRH6', 'ProteinLength': 189, 'germplasm_id': 5}, {'id': 6, 'Locus': 'AT2G03800', 'germplasm': 'gek1-1', 'phenotype': 'Ethanol hypersensitivity.', 'pubmed': 15215505, 'locusgene.id': 6, 'locusgene.Locus': 'AT2G03800', 'Gene': 'GEK1', 'ProteinLength': 196, 'germplasm_id': 6}, {'id': 7, 'Locus': 'AT2G04240', 'germplasm': 'xerico', 'phenotype': 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 'pubmed': 17933900, 'locusgene.id': 7, 'locusgene.Locus': 'AT2G04240', 'Gene': 'XERICO', 'ProteinLength': 256, 'germplasm_id': 7}, {'id': 8, 'Locus': 'AT2G05210', 'germplasm': 'pot1-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17627276, 'locusgene.id': 8, 'locusgene.Locus': 'AT2G05210', 'Gene': 'POT1A', 'ProteinLength': 221, 'germplasm_id': 8}, {'id': 9, 'Locus': 'AT3G02130', 'germplasm': 'rpk2-2', 'phenotype': 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 'pubmed': 17419837, 'locusgene.id': 9, 'locusgene.Locus': 'AT3G02130', 'Gene': 'RPK2', 'ProteinLength': 284, 'germplasm_id': 9}, {'id': 10, 'Locus': 'AT3G02140', 'germplasm': 'afp4-1', 'phenotype': 'Decreased germination on high concentrations of glucose and sorbitol.', 'pubmed': 18484180, 'locusgene.id': 10, 'locusgene.Locus': 'AT3G02140', 'Gene': 'TMAC2', 'ProteinLength': 300, 'germplasm_id': 10}, {'id': 11, 'Locus': 'AT3G02230', 'germplasm': 'rgp1-1', 'phenotype': 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 'pubmed': 21478444, 'locusgene.id': 11, 'locusgene.Locus': 'AT3G02230', 'Gene': 'RGP1', 'ProteinLength': 301, 'germplasm_id': 11}, {'id': 12, 'Locus': 'AT3G02260', 'germplasm': 'tir3-1 RGLG1:rglg1 rglg2', 'phenotype': 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 'pubmed': 17586653, 'locusgene.id': 12, 'locusgene.Locus': 'AT3G02260', 'Gene': 'BIG', 'ProteinLength': 279, 'germplasm_id': 12}, {'id': 13, 'Locus': 'AT3G02310', 'germplasm': 'sep2-1', 'phenotype': 'Non-described subtle phenotype.', 'pubmed': 10821278, 'locusgene.id': 13, 'locusgene.Locus': 'AT3G02310', 'Gene': 'SEP2', 'ProteinLength': 175, 'germplasm_id': 13}, {'id': 14, 'Locus': 'AT3G02680', 'germplasm': 'atnbs1-1', 'phenotype': 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 'pubmed': 17672843, 'locusgene.id': 14, 'locusgene.Locus': 'AT3G02680', 'Gene': 'NBS1', 'ProteinLength': 190, 'germplasm_id': 14}, {'id': 15, 'Locus': 'AT3G02850', 'germplasm': 'CS3816', 'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'pubmed': 17568770, 'locusgene.id': 15, 'locusgene.Locus': 'AT3G02850', 'Gene': 'SKOR', 'ProteinLength': 234, 'germplasm_id': 15}, {'id': 16, 'Locus': 'AT3G02870', 'germplasm': 'vtc4-1', 'phenotype': 'ascorbate deficient', 'pubmed': 16595667, 'locusgene.id': 16, 'locusgene.Locus': 'AT3G02870', 'Gene': 'VTC4', 'ProteinLength': 311, 'germplasm_id': 16}, {'id': 17, 'Locus': 'AT3G03260', 'germplasm': 'hdg8-1', 'phenotype': 'No visible phenotype.', 'pubmed': 16778018, 'locusgene.id': 17, 'locusgene.Locus': 'AT3G03260', 'Gene': 'HDG8', 'ProteinLength': 194, 'germplasm_id': 17}, {'id': 18, 'Locus': 'AT4G14790', 'germplasm': 'pdd17', 'phenotype': 'Defective pollen development.', 'pubmed': 19237690, 'locusgene.id': 18, 'locusgene.Locus': 'AT4G14790', 'Gene': 'SUV3', 'ProteinLength': 312, 'germplasm_id': 18}, {'id': 19, 'Locus': 'AT4G15210', 'germplasm': 'bmy1-2', 'phenotype': 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 'pubmed': 16297066, 'locusgene.id': 19, 'locusgene.Locus': 'AT4G15210', 'Gene': 'BAM5', 'ProteinLength': 313, 'germplasm_id': 19}, {'id': 20, 'Locus': 'AT4G15560', 'germplasm': 'cla1-1', 'phenotype': 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 'pubmed': 10982425, 'locusgene.id': 20, 'locusgene.Locus': 'AT4G15560', 'Gene': 'DXS', 'ProteinLength': 219, 'germplasm_id': 20}, {'id': 21, 'Locus': 'AT4G15570', 'germplasm': 'maa3', 'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'pubmed': 18772186, 'locusgene.id': 21, 'locusgene.Locus': 'AT4G15570', 'Gene': 'MAA3', 'ProteinLength': 294, 'germplasm_id': 21}, {'id': 22, 'Locus': 'AT4G15802', 'germplasm': 'Athspb-2', 'phenotype': 'Early flowering, reduced fertility, aborted seeds.', 'pubmed': 20388662, 'locusgene.id': 22, 'locusgene.Locus': 'AT4G15802', 'Gene': 'HSBP', 'ProteinLength': 254, 'germplasm_id': 22}, {'id': 23, 'Locus': 'AT4G15880', 'germplasm': 'esd4-2', 'phenotype': 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 'pubmed': 17513499, 'locusgene.id': 23, 'locusgene.Locus': 'AT4G15880', 'Gene': 'ESD4', 'ProteinLength': 265, 'germplasm_id': 23}, {'id': 24, 'Locus': 'AT4G16420', 'germplasm': 'prz1-1', 'phenotype': 'Altered response to auxin and cytokinin', 'pubmed': 12747832, 'locusgene.id': 24, 'locusgene.Locus': 'AT4G16420', 'Gene': 'ADA2B', 'ProteinLength': 279, 'germplasm_id': 24}, {'id': 25, 'Locus': 'AT4G16480', 'germplasm': 'atint4-2', 'phenotype': 'No visible phenotype.', 'pubmed': 16603666, 'locusgene.id': 25, 'locusgene.Locus': 'AT4G16480', 'Gene': 'INT4', 'ProteinLength': 284, 'germplasm_id': 25}, {'id': 26, 'Locus': 'AT5G10480', 'germplasm': 'pas2-3', 'phenotype': 'Segregates 25% embryo lethal.', 'pubmed': 18799749, 'locusgene.id': 26, 'locusgene.Locus': 'AT5G10480', 'Gene': 'PAS2', 'ProteinLength': 301, 'germplasm_id': 26}, {'id': 27, 'Locus': 'AT5G10510', 'germplasm': 'plt3-1', 'phenotype': 'Short roots and shortened root meristem.', 'pubmed': 17960244, 'locusgene.id': 27, 'locusgene.Locus': 'AT5G10510', 'Gene': 'AIL6', 'ProteinLength': 310, 'germplasm_id': 27}, {'id': 28, 'Locus': 'AT5G11110', 'germplasm': 'kns2', 'phenotype': 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 'pubmed': 18779216, 'locusgene.id': 28, 'locusgene.Locus': 'AT5G11110', 'Gene': 'SPS2', 'ProteinLength': 232, 'germplasm_id': 28}, {'id': 29, 'Locus': 'AT5G11260', 'germplasm': 'hy5-101', 'phenotype': 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 'pubmed': 16891401, 'locusgene.id': 29, 'locusgene.Locus': 'AT5G11260', 'Gene': 'HY5', 'ProteinLength': 221, 'germplasm_id': 29}, {'id': 30, 'Locus': 'AT5G11510', 'germplasm': 'myb3r4-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17287251, 'locusgene.id': 30, 'locusgene.Locus': 'AT5G11510', 'Gene': 'MYB3R-4', 'ProteinLength': 336, 'germplasm_id': 30}, {'id': 31, 'Locus': 'AT5G12200', 'germplasm': 'pyd2-2', 'phenotype': 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 'pubmed': 19413687, 'locusgene.id': 31, 'locusgene.Locus': 'AT5G12200', 'Gene': 'PYD2', 'ProteinLength': 310, 'germplasm_id': 31}, {'id': 32, 'Locus': 'AT5G13290', 'germplasm': 'crn-1', 'phenotype': 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 'pubmed': 12345678, 'locusgene.id': 32, 'locusgene.Locus': 'AT5G13290', 'Gene': 'CRN', 'ProteinLength': 189, 'germplasm_id': 32}]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "import csv \n",
    "import re\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Exam',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "            sql = \"\"\"SELECT * FROM germplasm INNER JOIN locusgene ON germplasm.id = locusgene.germplasm_id AND germplasm.Locus = locusgene.Locus;\"\"\"\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()\n",
    "           \n",
    "            with open(\"ExamReports.txt\", \"w\") as answers:\n",
    "                answers.write(str(results)+\"\\n\")  # the write function, using \\n (newline)\n",
    "                answers.close()\n",
    "\n",
    "            connection.commit() #esto solo si esta la opcion autocommit false \n",
    "        \n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()\n",
    "\n",
    "answers.close()\n",
    "\n",
    "checkcontent = open(\"ExamReports.txt\", \"r\")\n",
    "print(checkcontent.read()) # print the content of the file\n",
    "checkcontent.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Query 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[{'id': 1, 'Locus': 'AT1G01040', 'germplasm': 'CS3828', 'phenotype': 'Increased abundance of miRNA precursors.', 'pubmed': 17369351, 'locusgene.id': 1, 'locusgene.Locus': 'AT1G01040', 'Gene': 'DCL1', 'ProteinLength': 332, 'germplasm_id': 1}, {'id': 2, 'Locus': 'AT1G01060', 'germplasm': 'lhy-101', 'phenotype': 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 'pubmed': 16891401, 'locusgene.id': 2, 'locusgene.Locus': 'AT1G01060', 'Gene': 'LHY', 'ProteinLength': 290, 'germplasm_id': 2}, {'id': 3, 'Locus': 'AT1G01140', 'germplasm': 'SALK_058629', 'phenotype': 'hypersensitive to low potassium media', 'pubmed': 17486125, 'locusgene.id': 3, 'locusgene.Locus': 'AT1G01140', 'Gene': 'CIPK9', 'ProteinLength': 223, 'germplasm_id': 3}, {'id': 4, 'Locus': 'AT1G01220', 'germplasm': 'SALK_012400C', 'phenotype': 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 'pubmed': 18199744, 'locusgene.id': 4, 'locusgene.Locus': 'AT1G01220', 'Gene': 'FKGP', 'ProteinLength': 190, 'germplasm_id': 4}, {'id': 5, 'Locus': 'AT2G03720', 'germplasm': 'SALK_042433', 'phenotype': 'Multiple straight hairs', 'pubmed': 16367956, 'locusgene.id': 5, 'locusgene.Locus': 'AT2G03720', 'Gene': 'MRH6', 'ProteinLength': 189, 'germplasm_id': 5}, {'id': 6, 'Locus': 'AT2G03800', 'germplasm': 'gek1-1', 'phenotype': 'Ethanol hypersensitivity.', 'pubmed': 15215505, 'locusgene.id': 6, 'locusgene.Locus': 'AT2G03800', 'Gene': 'GEK1', 'ProteinLength': 196, 'germplasm_id': 6}, {'id': 7, 'Locus': 'AT2G04240', 'germplasm': 'xerico', 'phenotype': 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 'pubmed': 17933900, 'locusgene.id': 7, 'locusgene.Locus': 'AT2G04240', 'Gene': 'XERICO', 'ProteinLength': 256, 'germplasm_id': 7}, {'id': 8, 'Locus': 'AT2G05210', 'germplasm': 'pot1-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17627276, 'locusgene.id': 8, 'locusgene.Locus': 'AT2G05210', 'Gene': 'POT1A', 'ProteinLength': 221, 'germplasm_id': 8}, {'id': 9, 'Locus': 'AT3G02130', 'germplasm': 'rpk2-2', 'phenotype': 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 'pubmed': 17419837, 'locusgene.id': 9, 'locusgene.Locus': 'AT3G02130', 'Gene': 'RPK2', 'ProteinLength': 284, 'germplasm_id': 9}, {'id': 10, 'Locus': 'AT3G02140', 'germplasm': 'afp4-1', 'phenotype': 'Decreased germination on high concentrations of glucose and sorbitol.', 'pubmed': 18484180, 'locusgene.id': 10, 'locusgene.Locus': 'AT3G02140', 'Gene': 'TMAC2', 'ProteinLength': 300, 'germplasm_id': 10}, {'id': 11, 'Locus': 'AT3G02230', 'germplasm': 'rgp1-1', 'phenotype': 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 'pubmed': 21478444, 'locusgene.id': 11, 'locusgene.Locus': 'AT3G02230', 'Gene': 'RGP1', 'ProteinLength': 301, 'germplasm_id': 11}, {'id': 12, 'Locus': 'AT3G02260', 'germplasm': 'tir3-1 RGLG1:rglg1 rglg2', 'phenotype': 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 'pubmed': 17586653, 'locusgene.id': 12, 'locusgene.Locus': 'AT3G02260', 'Gene': 'BIG', 'ProteinLength': 279, 'germplasm_id': 12}, {'id': 13, 'Locus': 'AT3G02310', 'germplasm': 'sep2-1', 'phenotype': 'Non-described subtle phenotype.', 'pubmed': 10821278, 'locusgene.id': 13, 'locusgene.Locus': 'AT3G02310', 'Gene': 'SEP2', 'ProteinLength': 175, 'germplasm_id': 13}, {'id': 14, 'Locus': 'AT3G02680', 'germplasm': 'atnbs1-1', 'phenotype': 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 'pubmed': 17672843, 'locusgene.id': 14, 'locusgene.Locus': 'AT3G02680', 'Gene': 'NBS1', 'ProteinLength': 190, 'germplasm_id': 14}, {'id': 15, 'Locus': 'AT3G02850', 'germplasm': 'CS3816', 'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'pubmed': 17568770, 'locusgene.id': 15, 'locusgene.Locus': 'AT3G02850', 'Gene': 'SKOR', 'ProteinLength': 234, 'germplasm_id': 15}, {'id': 16, 'Locus': 'AT3G02870', 'germplasm': 'vtc4-1', 'phenotype': 'ascorbate deficient', 'pubmed': 16595667, 'locusgene.id': 16, 'locusgene.Locus': 'AT3G02870', 'Gene': 'VTC4', 'ProteinLength': 311, 'germplasm_id': 16}, {'id': 17, 'Locus': 'AT3G03260', 'germplasm': 'hdg8-1', 'phenotype': 'No visible phenotype.', 'pubmed': 16778018, 'locusgene.id': 17, 'locusgene.Locus': 'AT3G03260', 'Gene': 'HDG8', 'ProteinLength': 194, 'germplasm_id': 17}, {'id': 18, 'Locus': 'AT4G14790', 'germplasm': 'pdd17', 'phenotype': 'Defective pollen development.', 'pubmed': 19237690, 'locusgene.id': 18, 'locusgene.Locus': 'AT4G14790', 'Gene': 'SUV3', 'ProteinLength': 312, 'germplasm_id': 18}, {'id': 19, 'Locus': 'AT4G15210', 'germplasm': 'bmy1-2', 'phenotype': 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 'pubmed': 16297066, 'locusgene.id': 19, 'locusgene.Locus': 'AT4G15210', 'Gene': 'BAM5', 'ProteinLength': 313, 'germplasm_id': 19}, {'id': 20, 'Locus': 'AT4G15560', 'germplasm': 'cla1-1', 'phenotype': 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 'pubmed': 10982425, 'locusgene.id': 20, 'locusgene.Locus': 'AT4G15560', 'Gene': 'DXS', 'ProteinLength': 219, 'germplasm_id': 20}, {'id': 21, 'Locus': 'AT4G15570', 'germplasm': 'maa3', 'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'pubmed': 18772186, 'locusgene.id': 21, 'locusgene.Locus': 'AT4G15570', 'Gene': 'MAA3', 'ProteinLength': 294, 'germplasm_id': 21}, {'id': 22, 'Locus': 'AT4G15802', 'germplasm': 'Athspb-2', 'phenotype': 'Early flowering, reduced fertility, aborted seeds.', 'pubmed': 20388662, 'locusgene.id': 22, 'locusgene.Locus': 'AT4G15802', 'Gene': 'HSBP', 'ProteinLength': 254, 'germplasm_id': 22}, {'id': 23, 'Locus': 'AT4G15880', 'germplasm': 'esd4-2', 'phenotype': 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 'pubmed': 17513499, 'locusgene.id': 23, 'locusgene.Locus': 'AT4G15880', 'Gene': 'ESD4', 'ProteinLength': 265, 'germplasm_id': 23}, {'id': 24, 'Locus': 'AT4G16420', 'germplasm': 'prz1-1', 'phenotype': 'Altered response to auxin and cytokinin', 'pubmed': 12747832, 'locusgene.id': 24, 'locusgene.Locus': 'AT4G16420', 'Gene': 'ADA2B', 'ProteinLength': 279, 'germplasm_id': 24}, {'id': 25, 'Locus': 'AT4G16480', 'germplasm': 'atint4-2', 'phenotype': 'No visible phenotype.', 'pubmed': 16603666, 'locusgene.id': 25, 'locusgene.Locus': 'AT4G16480', 'Gene': 'INT4', 'ProteinLength': 284, 'germplasm_id': 25}, {'id': 26, 'Locus': 'AT5G10480', 'germplasm': 'pas2-3', 'phenotype': 'Segregates 25% embryo lethal.', 'pubmed': 18799749, 'locusgene.id': 26, 'locusgene.Locus': 'AT5G10480', 'Gene': 'PAS2', 'ProteinLength': 301, 'germplasm_id': 26}, {'id': 27, 'Locus': 'AT5G10510', 'germplasm': 'plt3-1', 'phenotype': 'Short roots and shortened root meristem.', 'pubmed': 17960244, 'locusgene.id': 27, 'locusgene.Locus': 'AT5G10510', 'Gene': 'AIL6', 'ProteinLength': 310, 'germplasm_id': 27}, {'id': 28, 'Locus': 'AT5G11110', 'germplasm': 'kns2', 'phenotype': 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 'pubmed': 18779216, 'locusgene.id': 28, 'locusgene.Locus': 'AT5G11110', 'Gene': 'SPS2', 'ProteinLength': 232, 'germplasm_id': 28}, {'id': 29, 'Locus': 'AT5G11260', 'germplasm': 'hy5-101', 'phenotype': 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 'pubmed': 16891401, 'locusgene.id': 29, 'locusgene.Locus': 'AT5G11260', 'Gene': 'HY5', 'ProteinLength': 221, 'germplasm_id': 29}, {'id': 30, 'Locus': 'AT5G11510', 'germplasm': 'myb3r4-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17287251, 'locusgene.id': 30, 'locusgene.Locus': 'AT5G11510', 'Gene': 'MYB3R-4', 'ProteinLength': 336, 'germplasm_id': 30}, {'id': 31, 'Locus': 'AT5G12200', 'germplasm': 'pyd2-2', 'phenotype': 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 'pubmed': 19413687, 'locusgene.id': 31, 'locusgene.Locus': 'AT5G12200', 'Gene': 'PYD2', 'ProteinLength': 310, 'germplasm_id': 31}, {'id': 32, 'Locus': 'AT5G13290', 'germplasm': 'crn-1', 'phenotype': 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 'pubmed': 12345678, 'locusgene.id': 32, 'locusgene.Locus': 'AT5G13290', 'Gene': 'CRN', 'ProteinLength': 189, 'germplasm_id': 32}]\n",
      "Question 2\n",
      "[{'id': 15, 'Locus': 'AT3G02850', 'Gene': 'SKOR', 'ProteinLength': 234, 'germplasm_id': 15, 'germplasm.id': 15, 'germplasm.Locus': 'AT3G02850', 'germplasm': 'CS3816', 'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'pubmed': 17568770}, {'id': 21, 'Locus': 'AT4G15570', 'Gene': 'MAA3', 'ProteinLength': 294, 'germplasm_id': 21, 'germplasm.id': 21, 'germplasm.Locus': 'AT4G15570', 'germplasm': 'maa3', 'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'pubmed': 18772186}]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "import csv \n",
    "import re\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Exam',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "            sql = \"\"\"SELECT * FROM locusgene INNER JOIN germplasm ON germplasm.id = locusgene.germplasm_id AND germplasm.Locus = locusgene.Locus WHERE Gene='SKOR'OR Gene='MAA3'\"\"\"\n",
    "            cursor.execute(sql)\n",
    "            results = cursor.fetchall()\n",
    "           \n",
    "            with open(\"ExamReports.txt\", \"a\") as answers:\n",
    "                answers.write( \"Question 2\\n\"+str(results)+\"\\n\")  # the write function, using \\n (newline)\n",
    "                answers.close()\n",
    "\n",
    "            connection.commit() #esto solo si esta la opcion autocommit false \n",
    "        \n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()\n",
    "\n",
    "answers.close()\n",
    "\n",
    "checkcontent = open(\"ExamReports.txt\", \"r\")\n",
    "print(checkcontent.read()) # print the content of the file\n",
    "checkcontent.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "1 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Number Of Matches</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>4</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[(4,)]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql SELECT COUNT(*) AS \"Number Of Matches\" FROM locusgene INNER JOIN germplasm ON germplasm.id = locusgene.germplasm_id AND germplasm.Locus = locusgene.Locus WHERE germplasm.Locus LIKE 'AT1G%'; \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[{'id': 1, 'Locus': 'AT1G01040', 'germplasm': 'CS3828', 'phenotype': 'Increased abundance of miRNA precursors.', 'pubmed': 17369351, 'locusgene.id': 1, 'locusgene.Locus': 'AT1G01040', 'Gene': 'DCL1', 'ProteinLength': 332, 'germplasm_id': 1}, {'id': 2, 'Locus': 'AT1G01060', 'germplasm': 'lhy-101', 'phenotype': 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 'pubmed': 16891401, 'locusgene.id': 2, 'locusgene.Locus': 'AT1G01060', 'Gene': 'LHY', 'ProteinLength': 290, 'germplasm_id': 2}, {'id': 3, 'Locus': 'AT1G01140', 'germplasm': 'SALK_058629', 'phenotype': 'hypersensitive to low potassium media', 'pubmed': 17486125, 'locusgene.id': 3, 'locusgene.Locus': 'AT1G01140', 'Gene': 'CIPK9', 'ProteinLength': 223, 'germplasm_id': 3}, {'id': 4, 'Locus': 'AT1G01220', 'germplasm': 'SALK_012400C', 'phenotype': 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 'pubmed': 18199744, 'locusgene.id': 4, 'locusgene.Locus': 'AT1G01220', 'Gene': 'FKGP', 'ProteinLength': 190, 'germplasm_id': 4}, {'id': 5, 'Locus': 'AT2G03720', 'germplasm': 'SALK_042433', 'phenotype': 'Multiple straight hairs', 'pubmed': 16367956, 'locusgene.id': 5, 'locusgene.Locus': 'AT2G03720', 'Gene': 'MRH6', 'ProteinLength': 189, 'germplasm_id': 5}, {'id': 6, 'Locus': 'AT2G03800', 'germplasm': 'gek1-1', 'phenotype': 'Ethanol hypersensitivity.', 'pubmed': 15215505, 'locusgene.id': 6, 'locusgene.Locus': 'AT2G03800', 'Gene': 'GEK1', 'ProteinLength': 196, 'germplasm_id': 6}, {'id': 7, 'Locus': 'AT2G04240', 'germplasm': 'xerico', 'phenotype': 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 'pubmed': 17933900, 'locusgene.id': 7, 'locusgene.Locus': 'AT2G04240', 'Gene': 'XERICO', 'ProteinLength': 256, 'germplasm_id': 7}, {'id': 8, 'Locus': 'AT2G05210', 'germplasm': 'pot1-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17627276, 'locusgene.id': 8, 'locusgene.Locus': 'AT2G05210', 'Gene': 'POT1A', 'ProteinLength': 221, 'germplasm_id': 8}, {'id': 9, 'Locus': 'AT3G02130', 'germplasm': 'rpk2-2', 'phenotype': 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 'pubmed': 17419837, 'locusgene.id': 9, 'locusgene.Locus': 'AT3G02130', 'Gene': 'RPK2', 'ProteinLength': 284, 'germplasm_id': 9}, {'id': 10, 'Locus': 'AT3G02140', 'germplasm': 'afp4-1', 'phenotype': 'Decreased germination on high concentrations of glucose and sorbitol.', 'pubmed': 18484180, 'locusgene.id': 10, 'locusgene.Locus': 'AT3G02140', 'Gene': 'TMAC2', 'ProteinLength': 300, 'germplasm_id': 10}, {'id': 11, 'Locus': 'AT3G02230', 'germplasm': 'rgp1-1', 'phenotype': 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 'pubmed': 21478444, 'locusgene.id': 11, 'locusgene.Locus': 'AT3G02230', 'Gene': 'RGP1', 'ProteinLength': 301, 'germplasm_id': 11}, {'id': 12, 'Locus': 'AT3G02260', 'germplasm': 'tir3-1 RGLG1:rglg1 rglg2', 'phenotype': 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 'pubmed': 17586653, 'locusgene.id': 12, 'locusgene.Locus': 'AT3G02260', 'Gene': 'BIG', 'ProteinLength': 279, 'germplasm_id': 12}, {'id': 13, 'Locus': 'AT3G02310', 'germplasm': 'sep2-1', 'phenotype': 'Non-described subtle phenotype.', 'pubmed': 10821278, 'locusgene.id': 13, 'locusgene.Locus': 'AT3G02310', 'Gene': 'SEP2', 'ProteinLength': 175, 'germplasm_id': 13}, {'id': 14, 'Locus': 'AT3G02680', 'germplasm': 'atnbs1-1', 'phenotype': 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 'pubmed': 17672843, 'locusgene.id': 14, 'locusgene.Locus': 'AT3G02680', 'Gene': 'NBS1', 'ProteinLength': 190, 'germplasm_id': 14}, {'id': 15, 'Locus': 'AT3G02850', 'germplasm': 'CS3816', 'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'pubmed': 17568770, 'locusgene.id': 15, 'locusgene.Locus': 'AT3G02850', 'Gene': 'SKOR', 'ProteinLength': 234, 'germplasm_id': 15}, {'id': 16, 'Locus': 'AT3G02870', 'germplasm': 'vtc4-1', 'phenotype': 'ascorbate deficient', 'pubmed': 16595667, 'locusgene.id': 16, 'locusgene.Locus': 'AT3G02870', 'Gene': 'VTC4', 'ProteinLength': 311, 'germplasm_id': 16}, {'id': 17, 'Locus': 'AT3G03260', 'germplasm': 'hdg8-1', 'phenotype': 'No visible phenotype.', 'pubmed': 16778018, 'locusgene.id': 17, 'locusgene.Locus': 'AT3G03260', 'Gene': 'HDG8', 'ProteinLength': 194, 'germplasm_id': 17}, {'id': 18, 'Locus': 'AT4G14790', 'germplasm': 'pdd17', 'phenotype': 'Defective pollen development.', 'pubmed': 19237690, 'locusgene.id': 18, 'locusgene.Locus': 'AT4G14790', 'Gene': 'SUV3', 'ProteinLength': 312, 'germplasm_id': 18}, {'id': 19, 'Locus': 'AT4G15210', 'germplasm': 'bmy1-2', 'phenotype': 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 'pubmed': 16297066, 'locusgene.id': 19, 'locusgene.Locus': 'AT4G15210', 'Gene': 'BAM5', 'ProteinLength': 313, 'germplasm_id': 19}, {'id': 20, 'Locus': 'AT4G15560', 'germplasm': 'cla1-1', 'phenotype': 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 'pubmed': 10982425, 'locusgene.id': 20, 'locusgene.Locus': 'AT4G15560', 'Gene': 'DXS', 'ProteinLength': 219, 'germplasm_id': 20}, {'id': 21, 'Locus': 'AT4G15570', 'germplasm': 'maa3', 'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'pubmed': 18772186, 'locusgene.id': 21, 'locusgene.Locus': 'AT4G15570', 'Gene': 'MAA3', 'ProteinLength': 294, 'germplasm_id': 21}, {'id': 22, 'Locus': 'AT4G15802', 'germplasm': 'Athspb-2', 'phenotype': 'Early flowering, reduced fertility, aborted seeds.', 'pubmed': 20388662, 'locusgene.id': 22, 'locusgene.Locus': 'AT4G15802', 'Gene': 'HSBP', 'ProteinLength': 254, 'germplasm_id': 22}, {'id': 23, 'Locus': 'AT4G15880', 'germplasm': 'esd4-2', 'phenotype': 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 'pubmed': 17513499, 'locusgene.id': 23, 'locusgene.Locus': 'AT4G15880', 'Gene': 'ESD4', 'ProteinLength': 265, 'germplasm_id': 23}, {'id': 24, 'Locus': 'AT4G16420', 'germplasm': 'prz1-1', 'phenotype': 'Altered response to auxin and cytokinin', 'pubmed': 12747832, 'locusgene.id': 24, 'locusgene.Locus': 'AT4G16420', 'Gene': 'ADA2B', 'ProteinLength': 279, 'germplasm_id': 24}, {'id': 25, 'Locus': 'AT4G16480', 'germplasm': 'atint4-2', 'phenotype': 'No visible phenotype.', 'pubmed': 16603666, 'locusgene.id': 25, 'locusgene.Locus': 'AT4G16480', 'Gene': 'INT4', 'ProteinLength': 284, 'germplasm_id': 25}, {'id': 26, 'Locus': 'AT5G10480', 'germplasm': 'pas2-3', 'phenotype': 'Segregates 25% embryo lethal.', 'pubmed': 18799749, 'locusgene.id': 26, 'locusgene.Locus': 'AT5G10480', 'Gene': 'PAS2', 'ProteinLength': 301, 'germplasm_id': 26}, {'id': 27, 'Locus': 'AT5G10510', 'germplasm': 'plt3-1', 'phenotype': 'Short roots and shortened root meristem.', 'pubmed': 17960244, 'locusgene.id': 27, 'locusgene.Locus': 'AT5G10510', 'Gene': 'AIL6', 'ProteinLength': 310, 'germplasm_id': 27}, {'id': 28, 'Locus': 'AT5G11110', 'germplasm': 'kns2', 'phenotype': 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 'pubmed': 18779216, 'locusgene.id': 28, 'locusgene.Locus': 'AT5G11110', 'Gene': 'SPS2', 'ProteinLength': 232, 'germplasm_id': 28}, {'id': 29, 'Locus': 'AT5G11260', 'germplasm': 'hy5-101', 'phenotype': 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 'pubmed': 16891401, 'locusgene.id': 29, 'locusgene.Locus': 'AT5G11260', 'Gene': 'HY5', 'ProteinLength': 221, 'germplasm_id': 29}, {'id': 30, 'Locus': 'AT5G11510', 'germplasm': 'myb3r4-1', 'phenotype': 'No visible phenotype.', 'pubmed': 17287251, 'locusgene.id': 30, 'locusgene.Locus': 'AT5G11510', 'Gene': 'MYB3R-4', 'ProteinLength': 336, 'germplasm_id': 30}, {'id': 31, 'Locus': 'AT5G12200', 'germplasm': 'pyd2-2', 'phenotype': 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 'pubmed': 19413687, 'locusgene.id': 31, 'locusgene.Locus': 'AT5G12200', 'Gene': 'PYD2', 'ProteinLength': 310, 'germplasm_id': 31}, {'id': 32, 'Locus': 'AT5G13290', 'germplasm': 'crn-1', 'phenotype': 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 'pubmed': 12345678, 'locusgene.id': 32, 'locusgene.Locus': 'AT5G13290', 'Gene': 'CRN', 'ProteinLength': 189, 'germplasm_id': 32}]\n",
      "Question 2\n",
      "[{'id': 15, 'Locus': 'AT3G02850', 'Gene': 'SKOR', 'ProteinLength': 234, 'germplasm_id': 15, 'germplasm.id': 15, 'germplasm.Locus': 'AT3G02850', 'germplasm': 'CS3816', 'phenotype': 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 'pubmed': 17568770}, {'id': 21, 'Locus': 'AT4G15570', 'Gene': 'MAA3', 'ProteinLength': 294, 'germplasm_id': 21, 'germplasm.id': 21, 'germplasm.Locus': 'AT4G15570', 'germplasm': 'maa3', 'phenotype': 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 'pubmed': 18772186}]\n",
      "Question 3\n",
      "Counts for Chromosome 1\n",
      "[{'Number Of Matches': 4}]\n",
      "Counts for Chromosome 2\n",
      "[{'Number Of Matches': 4}]\n",
      "Counts for Chromosome 3\n",
      "[{'Number Of Matches': 9}]\n",
      "Counts for Chromosome 4\n",
      "[{'Number Of Matches': 4}]\n",
      "Counts for Chromosome 5\n",
      "[{'Number Of Matches': 7}]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pymysql.cursors\n",
    "import csv \n",
    "import re\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Exam',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "            sql1 = \"\"\"SELECT COUNT(*) AS \"Number Of Matches\" FROM locusgene INNER JOIN germplasm ON germplasm.id = locusgene.germplasm_id AND germplasm.Locus = locusgene.Locus WHERE germplasm.Locus LIKE 'AT1G%'\"\"\"\n",
    "            cursor.execute(sql1)\n",
    "            results1 = cursor.fetchall()\n",
    "            sql2 = \"\"\"SELECT COUNT(*) AS \"Number Of Matches\" FROM locusgene INNER JOIN germplasm ON germplasm.id = locusgene.germplasm_id AND germplasm.Locus = locusgene.Locus WHERE germplasm.Locus LIKE 'AT2G%'\"\"\"\n",
    "            cursor.execute(sql2)\n",
    "            results2 = cursor.fetchall()\n",
    "            sql3 = \"\"\"SELECT COUNT(*) AS \"Number Of Matches\" FROM locusgene INNER JOIN germplasm ON germplasm.id = locusgene.germplasm_id AND germplasm.Locus = locusgene.Locus WHERE germplasm.Locus LIKE 'AT3G%'\"\"\"\n",
    "            cursor.execute(sql3)\n",
    "            results3 = cursor.fetchall()\n",
    "            sql4 = \"\"\"SELECT COUNT(*) AS \"Number Of Matches\" FROM locusgene INNER JOIN germplasm ON germplasm.id = locusgene.germplasm_id AND germplasm.Locus = locusgene.Locus WHERE germplasm.Locus LIKE 'AT4G%'\"\"\"\n",
    "            cursor.execute(sql4)\n",
    "            results4 = cursor.fetchall()\n",
    "            sql5 = \"\"\"SELECT COUNT(*) AS \"Number Of Matches\" FROM locusgene INNER JOIN germplasm ON germplasm.id = locusgene.germplasm_id AND germplasm.Locus = locusgene.Locus WHERE germplasm.Locus LIKE 'AT5G%'\"\"\"\n",
    "            cursor.execute(sql5)\n",
    "            results5 = cursor.fetchall()\n",
    "           \n",
    "            with open(\"ExamReports.txt\", \"a\") as answers:\n",
    "                answers.write( \"Question 3\\n\"+ \"Counts for Chromosome 1\\n\"+str(results1)+\"\\n\"+\"Counts for Chromosome 2\\n\"+str(results2)+\"\\n\"+\"Counts for Chromosome 3\\n\"+str(results3)+\"\\n\"+\"Counts for Chromosome 4\\n\"+str(results1)+\"\\n\"+\"Counts for Chromosome 5\\n\"+str(results5)+\"\\n\")  # the write function, using \\n (newline)\n",
    "                answers.close()\n",
    "\n",
    "            connection.commit() #esto solo si esta la opcion autocommit false \n",
    "        \n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()\n",
    "\n",
    "answers.close()\n",
    "\n",
    "checkcontent = open(\"ExamReports.txt\", \"r\")\n",
    "print(checkcontent.read()) # print the content of the file\n",
    "checkcontent.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
