{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Files\n",
    "\n",
    "Germplasm.tsv and LocusGene.tsv contain the datasets we need for the exam.\n",
    "\n",
    "Our objective is to create a database to contain the data in these files, insert the data into the database, then query the database in a variety of ways.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 1: Controls\n",
    "\n",
    " ##### Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx). (hint: This will help you decide how to load the data into the database)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This question asks to check if both files have the same data sequence based on their AGI Locus Codes. With this objective, I have made a script that firstly opens both files in order to read them, then It reads the info of each Locus colum row (for row in \"file\": mo1 = row[\"locus\"]), these colums contain the AGI Locus Codes. Finally if there is the same data in both objects It will print \"Both colums hace the same data\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv  #First of all we have to import the python functions that will be used \n",
    "import re\n",
    "with open(\"Germplasm.tsv\", \"r\") as germplasmfile,\\\n",
    "     open(\"LocusGene.tsv\", \"r\") as locusgenefile:\n",
    "        germplasm = csv.DictReader(germplasmfile, delimiter=\"\\t\", quotechar='\"') #Read the files, (with \\t delimiter as they are .tsv files)\n",
    "        locusgene = csv.DictReader(locusgenefile, delimiter=\"\\t\", quotechar='\"')\n",
    "        for row in germplasm:   #we ask to read each row of \"Locus\" column in germplasm file\n",
    "            mo1 = row[\"Locus\"]\n",
    "            #print ('Germplasm locus', mo1)\n",
    "        for row in locusgene:  #we ask to read each row \"Locus\" column in LocusGene file\n",
    "            mo2 = row[\"Locus\"]\n",
    "            #print ('Locus Gene locus', mo2)\n",
    "        if mo1 == mo2:         # If both colums have the same AGI Locus codes then print both rows have the same data\n",
    "                print (\"Both colums have the same data\")\n",
    "        else:\n",
    "               print (\"No match!!\")  # If rows un files are different then print no match!!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I also wanted to try another way to perform thos control, for this reason I have made another script. In this case I also open both files in order to read then but when I read each file I use the function next(). This function returns the next item in an iterator so It would read each line of both files. Then I set the condition that if row \"locus\" of each reading have the same data, then \"Both rows have the same data\" will be printed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv     #First of all we have to import the python functions that will be used \n",
    "import re\n",
    "with open(\"Germplasm.tsv\", \"r\") as germplasmfile,\\\n",
    "     open(\"LocusGene.tsv\", \"r\") as locusgenefile:\n",
    "        germplasm = csv.DictReader(germplasmfile, delimiter=\"\\t\", quotechar='\"') #Dictionary reader\n",
    "        row1 = next(germplasm)  #It will return the next item in an iterator, in this case in germplasm file \n",
    "        locusgene = csv.DictReader(locusgenefile, delimiter=\"\\t\", quotechar='\"')\n",
    "        row2 = next(locusgene)  ##It will return the next item in an iterator, in this case in locusgene file \n",
    "        if row1[\"Locus\"] == row2[\"Locus\"]:  # # If both locus rows for each file contains the same info, then print both rows have the same data\n",
    "                print (\"Both colums have the same data\")\n",
    "        else:\n",
    "               print (\"No match!!\") # If rows un files are different then print no match!!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem 2:  Design and create the database.  \n",
    "* It should have two tables - one for each of the two data files.\n",
    "* The two tables should be linked in a 1:1 relationship\n",
    "* you may use either sqlMagic or pymysql to build the database\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First of all, I started mySQL server (docker start course-mysql) and executed the connection lines in this Jupyter Notebook:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext sql\n",
    "#%config SqlMagic.autocommit=False\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql\n",
    "%sql show databases; #consult existing databases"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following commands are set only in case I need to delete the database to start again if I make any mistakes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%sql drop database Exam \n",
    "#%sql show databases;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once SQL is connected and the existing databases have been consulted, I proceed to create a new database called Exam ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql create database Exam;\n",
    "%sql show databases\n",
    "%sql use Exam; #This command tells the server which database I want to interact with"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The next step is to create the two tables that will contain the information corresponding to each one or the .tsv files. For each column data it is neccesary to set what kind of data It will contain (colum definition). In this case for both tables, the Primary Key will be the id (which is also an auto incremenet intefgrer), the rest of the colums contais data such as Vachar() or integres depending on the type of information they must contain.   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql CREATE TABLE germplasm(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, Locus VARCHAR(25) NOT NULL, germplasm VARCHAR(30) NOT NULL, phenotype VARCHAR(2000) NOT NULL, pubmed INTEGER NOT NULL);\n",
    "%sql DESCRIBE germplasm\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql CREATE TABLE locusgene(id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY, Locus VARCHAR(25) NOT NULL, Gene VARCHAR(20) NOT NULL, ProteinLength INTEGER NOT NULL, germplasm_id INTEGER NOT NULL);\n",
    "%sql DESCRIBE locusgene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem 3: Fill the database\n",
    "##### Using pymysql, create a Python script that reads the data from these files, and fills the database.  There are a variety of strategies to accomplish this.  I will give all strategies equal credit - do whichever one you are most confident with."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this exercise, first of all I have to connect pymysql with the database that is being used (in this case Exam). All the conection to database process is defined as 'connection' in order to optimze the script. Then I have read the data from germplasm file and filled the database with them. For this, I have set a loop that insert into each database column its corresponding information from the columns of the file (INSERT INTO \"rows in databade\" VALUES (\"{}\".format <--- Row[row name in tsv file]).\n",
    "\n",
    ".format is used to do variable substitutions and data formatting without having to specify the format of each data string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "import csv \n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Exam', #my database is Exam          \n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False # with this option I delay putting changes into the database until I write the connection.commit() function.\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open('Germplasm.tsv')as csvfile:\n",
    "            germplasm = csv.DictReader(csvfile, delimiter=\"\\t\", quotechar='\"' ) #with germplasm file open to read I made an insert from the file to the databasa for each row\n",
    "            for row in germplasm:\n",
    "                print(row)\n",
    "                sql = \"\"\"INSERT INTO germplasm (Locus, germplasm, phenotype, pubmed ) values \n",
    "                ('{}','{}','{}','{}')\"\"\".format (row[\"Locus\"], row[\"germplasm\"], row[\"phenotype\"], row[\"pubmed\"]) \n",
    "                cursor.execute(sql)\n",
    "        \n",
    "        connection.commit() #commit the inserts \n",
    "\n",
    "finally: \n",
    "    print (\"\")\n",
    "    connection.close() #If I close before I commit, the inserts are lost"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM germplasm #check that the information has been entered correctly"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now I have to introduce the information from locusgene.tsv into its correspondent database table. For this I have read the data from LocusGene.tsv file and filled the database with them but in this case it was necessary to establish the existing relationships between the two tables so in order to do this I have selected id data from germplasm table and then I have introduced the result of this selection as \"germplasmid\" into the germplasm_id column on the locusgene table row by row. The rest of the colums have been filled in the same way as in the previous table. \n",
    "\n",
    "I have also used te str.format() function as in the previous case. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "import csv \n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Exam', #my database is Exam\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        with open('LocusGene.tsv') as csvfile:\n",
    "            locusgene = csv.DictReader(csvfile, delimiter=\"\\t\", quotechar='\"' )\n",
    "            for row in locusgene:\n",
    "                print(row)\n",
    "                locus = row[\"Locus\"] \n",
    "                locus = locus.strip()\n",
    "                print(\"SELECT id FROM germplasm WHERE locus='{}'\".format(locus)) #select the id data for each row from germplasm\n",
    "                \n",
    "                sql = \"SELECT id FROM germplasm WHERE locus='{}'\".format(locus)\n",
    "                cursor.execute(sql) #executes the 'sql' slection\n",
    "                results = cursor.fetchall() #fetches all the rows of the query result\n",
    "                germplasmid = results[0]['id'] #from the query, select id data\n",
    "                print (germplasmid)\n",
    "                print (\"Project Id is {}\".format(locusgene))\n",
    "                sql = \"\"\"INSERT INTO locusgene (Locus, Gene, ProteinLength,germplasm_id) values \n",
    "                ('{}','{}','{}','{}')\"\"\".format (row[\"Locus\"], row[\"Gene\"], row[\"ProteinLength\"], germplasmid) #fill the table but in case of germplasm_id, fill the row with the result of the query done.\n",
    "                cursor.execute(sql)\n",
    "        \n",
    "        connection.commit()\n",
    "finally: \n",
    "    print (\"\")\n",
    "    connection.close() #If I close before I commit, the inserts are lost"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%sql SELECT * FROM locusgene #check that the information has been entered correctly"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem 4: Create reports, written to a file\n",
    "\n",
    "1. Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "\n",
    "2. Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "3. Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "4. Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "When creating reports 2 and 3, remember the \"Don't Repeat Yourself\" rule! \n",
    "\n",
    "All reports should be written to **the same file**.  You may name the file anything you wish."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Query 1\n",
    "\n",
    "In this query I have use pymysql in order to make an inner join wich contains the joined content of the two tables, that's why I have write \"germplasm.id = locusgene.germplasm_id AND germplasm.Locus = locusgene.Locus\" . \n",
    "\n",
    "Then I have writen the query result un ExamReports.txt. in this case, after open the file, I have used the write funcion 'w' because if I need to execute again the whole Problem 4 scripts, as this is the first one it will rewrite the file and the rest will be added after it.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "import csv \n",
    "import re\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Exam', #my database is Exam\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "            sql = \"\"\"SELECT * FROM germplasm INNER JOIN locusgene ON germplasm.id = locusgene.germplasm_id AND germplasm.Locus = locusgene.Locus;\"\"\"\n",
    "            cursor.execute(sql) #executes the 'sql' slection\n",
    "            results = cursor.fetchall() #fetches all the rows of the query result\n",
    "           \n",
    "            with open(\"ExamReports.txt\", \"w\") as answers:\n",
    "                answers.write(\"-->Question 1\\n\"+str(results)+\"\\n\")  # write query result in opened file. (\\n (newline)).\n",
    "                answers.close() #close file\n",
    "\n",
    "            connection.commit() #esto solo si esta la opcion autocommit false \n",
    "        \n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close() #close the connection as It is set connection.autocommit = False\n",
    "\n",
    "answers.close()\n",
    "\n",
    "checkcontent = open(\"ExamReports.txt\", \"r\")  # commands used to check if the file has been writen correctly\n",
    "print(checkcontent.read()) # print the content of the file\n",
    "checkcontent.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Query 2\n",
    "\n",
    "In this case I have selected the joined rows of both tables but only for the rows with SKOR and MAA3 un the locusgene.Gene column  \n",
    "\n",
    "Then I have added the query result in ExamReports.txt. In this case I have used de 'a' function.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "import csv \n",
    "import re\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Exam', #my database is Exam\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "            sql = \"\"\"SELECT * FROM locusgene INNER JOIN germplasm ON germplasm.id = locusgene.germplasm_id AND germplasm.Locus = locusgene.Locus WHERE Gene='SKOR'OR Gene='MAA3'\"\"\"\n",
    "            cursor.execute(sql) #executes the 'sql' slection\n",
    "            results = cursor.fetchall() #fetches all the rows of the query result\n",
    "            #Add the result to the ExamReports file\n",
    "            with open(\"ExamReports.txt\", \"a\") as answers:\n",
    "                answers.write( \"--> Question 2\\n\"+str(results)+\"\\n\")  # write query result in opened file. (\\n (newline)).\n",
    "                answers.close() #close file\n",
    "\n",
    "            connection.commit() #close the connection as It is set connection.autocommit = False\n",
    "\n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()\n",
    "\n",
    "answers.close()\n",
    "\n",
    "checkcontent = open(\"ExamReports.txt\", \"r\") # commands used to check if the file has been writen correctly\n",
    "print(checkcontent.read()) # print the content of the file\n",
    "checkcontent.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Query 3\n",
    "\n",
    "In this case I have use the COUNT function on order to count all the entries in joined tables but only in the rows WHERE the colum Locus of germplasm table was like 'AT1G%'. In other words, this command line counts all the entries for chromosme 1 in the joined table. \n",
    "\n",
    "Then I have repeated the same query for each chromosome, changing the 'ATxG%' part of the code (sorry for not complying with the Don'treapeatyourself rule) \n",
    "\n",
    "Then I have added the result of each query in ExamReports.txt. In this case I have used de 'a' function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "import csv \n",
    "import re\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Exam',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "            sql1 = \"\"\"SELECT COUNT(*) AS \"Number Of Matches\" FROM locusgene INNER JOIN germplasm ON germplasm.id = locusgene.germplasm_id AND germplasm.Locus = locusgene.Locus WHERE germplasm.Locus LIKE 'AT1G%'\"\"\"\n",
    "            cursor.execute(sql1)\n",
    "            results1 = cursor.fetchall()\n",
    "            sql2 = \"\"\"SELECT COUNT(*) AS \"Number Of Matches\" FROM locusgene INNER JOIN germplasm ON germplasm.id = locusgene.germplasm_id AND germplasm.Locus = locusgene.Locus WHERE germplasm.Locus LIKE 'AT2G%'\"\"\"\n",
    "            cursor.execute(sql2)\n",
    "            results2 = cursor.fetchall()\n",
    "            sql3 = \"\"\"SELECT COUNT(*) AS \"Number Of Matches\" FROM locusgene INNER JOIN germplasm ON germplasm.id = locusgene.germplasm_id AND germplasm.Locus = locusgene.Locus WHERE germplasm.Locus LIKE 'AT3G%'\"\"\"\n",
    "            cursor.execute(sql3)\n",
    "            results3 = cursor.fetchall()\n",
    "            sql4 = \"\"\"SELECT COUNT(*) AS \"Number Of Matches\" FROM locusgene INNER JOIN germplasm ON germplasm.id = locusgene.germplasm_id AND germplasm.Locus = locusgene.Locus WHERE germplasm.Locus LIKE 'AT4G%'\"\"\"\n",
    "            cursor.execute(sql4)\n",
    "            results4 = cursor.fetchall()\n",
    "            sql5 = \"\"\"SELECT COUNT(*) AS \"Number Of Matches\" FROM locusgene INNER JOIN germplasm ON germplasm.id = locusgene.germplasm_id AND germplasm.Locus = locusgene.Locus WHERE germplasm.Locus LIKE 'AT5G%'\"\"\"\n",
    "            cursor.execute(sql5)\n",
    "            results5 = cursor.fetchall()\n",
    "               #Add the diferent results to the ExamReports file\n",
    "            with open(\"ExamReports.txt\", \"a\") as answers:\n",
    "                answers.write( \"--> Question 3\\n\"+ \"Counts for Chromosome 1\\n\"+str(results1)+\"\\n\"+\"Counts for Chromosome 2\\n\"+str(results2)+\"\\n\"+\"Counts for Chromosome 3\\n\"+str(results3)+\"\\n\"+\"Counts for Chromosome 4\\n\"+str(results1)+\"\\n\"+\"Counts for Chromosome 5\\n\"+str(results5)+\"\\n\")  # write query result in opened file. (\\n (newline)).\n",
    "                answers.close() #close file\n",
    "\n",
    "            connection.commit() #close the connection as It is set connection.autocommit = False\n",
    "        \n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()\n",
    "\n",
    "answers.close()\n",
    "\n",
    "checkcontent = open(\"ExamReports.txt\", \"r\") # commands used to check if the file has been writen correctly\n",
    "print(checkcontent.read()) # print the content of the file\n",
    "checkcontent.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Query 4\n",
    "\n",
    "In this case I have use the AVG function in order to make an average of all the entries in ProteinLength column of locusgene table but only in the rows WHERE the Locus of germplasm table was like 'AT1G%'. This command line will make an average of all the entries in ProteinLenght row for chromosme 1 in the joined table.\n",
    "\n",
    "Then I have repeated the same query for each chromosome, changing the 'ATxG%' part of the code.\n",
    "\n",
    "Finally I have added the result of each query in ExamReports.txt. In this case I have used de 'a' function.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pymysql.cursors\n",
    "import csv \n",
    "import re\n",
    "\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Exam', #using Exams database \n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "connection.autocommit = False\n",
    "\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "            sql1 = \"\"\"SELECT AVG(locusgene.ProteinLength) AS \"Average protein length\" FROM locusgene INNER JOIN germplasm ON germplasm.id = locusgene.germplasm_id AND germplasm.Locus = locusgene.Locus WHERE germplasm.Locus LIKE 'AT1G%'\"\"\"\n",
    "            cursor.execute(sql1)\n",
    "            results1 = cursor.fetchall()\n",
    "            sql2 = \"\"\"SELECT AVG(locusgene.ProteinLength) AS \"Average protein length\" FROM locusgene INNER JOIN germplasm ON germplasm.id = locusgene.germplasm_id AND germplasm.Locus = locusgene.Locus WHERE germplasm.Locus LIKE 'AT2G%'\"\"\"\n",
    "            cursor.execute(sql2)\n",
    "            results2 = cursor.fetchall()\n",
    "            sql3 = \"\"\"SELECT AVG(locusgene.ProteinLength) AS \"Average protein length\" FROM locusgene INNER JOIN germplasm ON germplasm.id = locusgene.germplasm_id AND germplasm.Locus = locusgene.Locus WHERE germplasm.Locus LIKE 'AT3G%'\"\"\"\n",
    "            cursor.execute(sql3)\n",
    "            results3 = cursor.fetchall()\n",
    "            sql4 = \"\"\"SELECT AVG(locusgene.ProteinLength) AS \"Average protein length\" FROM locusgene INNER JOIN germplasm ON germplasm.id = locusgene.germplasm_id AND germplasm.Locus = locusgene.Locus WHERE germplasm.Locus LIKE 'AT4G%'\"\"\"\n",
    "            cursor.execute(sql4)\n",
    "            results4 = cursor.fetchall()\n",
    "            sql5 = \"\"\"SELECT AVG(locusgene.ProteinLength) AS \"Average protein length\" FROM locusgene INNER JOIN germplasm ON germplasm.id = locusgene.germplasm_id AND germplasm.Locus = locusgene.Locus WHERE germplasm.Locus LIKE 'AT5G%'\"\"\"\n",
    "            cursor.execute(sql5)\n",
    "            results5 = cursor.fetchall()\n",
    "               #Add the diferent results to the ExamReports file\n",
    "            with open(\"ExamReports.txt\", \"a\") as answers:\n",
    "                answers.write( \"--> Question 4\\n\"+ \"Average protein length for genes in Chromosome 1\\n\"+str(results1)+\"\\n\"+\"Average protein length for genes in Chromosome 2\\n\"+str(results2)+\"\\n\"+\"Average protein length for genes in Chromosome 3\\n\"+str(results3)+\"\\n\"+\"Average protein length for genes in Chromosome 4\\n\"+str(results1)+\"\\n\"+\"Average protein length for genes in Chromosome 5\\n\"+str(results5)+\"\\n\")  # write query result in opened file. (\\n (newline)).\n",
    "                answers.close() #close file\n",
    "\n",
    "            connection.commit() #close the connection as It is set connection.autocommit = False\n",
    "        \n",
    "finally:\n",
    "    print(\"\")\n",
    "    connection.close()\n",
    "\n",
    "answers.close()\n",
    "\n",
    "checkcontent = open(\"ExamReports.txt\", \"r\") # commands used to check if the file has been writen correctly\n",
    "print(checkcontent.read()) # print the content of the file\n",
    "checkcontent.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I know that these scripts are repetitive and not abstract, I have tried to optimize them but I have not been able to obtain the desired output. That is why I have decided to leave them in this simple and ugly way. I apologize, I am a novice in this field but I will keep practicing until I improve my programming skills and reach the expected level of a bioinformatics student. Even so, I know that it is probably not the expected level for this exam and that's why I think is necessary to apologize."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
